### DISCUSSION TITLE
Remove action in a splitAtom triggers unnecessary updates

### DISCUSSION THREAD

**Discussion by idrm** [pmndrs/jotai#2331] [General]
A `splitAtom` atom appears to trigger updates in all components that use atoms from the list that come after an atom being removed from the list. E.g. if the atom list consists of atom1, atom2, and atom3 (in that order), dispatching a remove action for atom2 will trigger an update in components using atom3 (those using atom1 are unaffected).

Here's a [CSB reproduction](https://codesandbox.io/p/sandbox/jotai-splitatom-unnecessary-rerender-5pt5ht?file=%2Fsrc%2FApp.tsx%3A10%2C5) of the issue (based on the example in the `splitAtom` docs).

One change from the original example is the  memoization of the `TodoItem` component. This is to prevent a `TodoList` re-render from triggering a new render in `TodoItem` instances that receive the same atom. `TodoItem` re-renders happen only due to the application of `useAtom` in them.

You can observe the behavior by opening the console and clicking on the 'remove' button of the second item: the console will show a re-render message for the third item (but not for the first item). If you tried removing the first them then a re-render occurs for both the second and third items.

I can't tell if this is the expected behavior or not. It can potentially affect performance when a large list is rendered, and the user removes an item near the start of the list.

**Comment by dai-shi** [comment]
I think you need to use `keyExtractor`, same idea as the JSX key.

**Comment by idrm** [comment]
I see that argument in the source code for `splitAtom` now; it isn't in the function's docs page.

If I understand the code correctly, each item's index in the list is used as the key when no `keyExtractor` is passed. Would it then be possible to add an optimization step that shifts those keys when the atom list is updated using the dispatch function returned by `splitAtom`, thus avoiding updates to the unaffected items/atoms? Calls to `setState` of the parent item list atom would still require a `keyExtractor` to avoid these unnecessary updates.

**Reply by dai-shi** [reply]
Yeah, we should improve the docs. It would be really appreciated if you can submit a PR to update the docs.

(I wonder if one of our examples with `splitAtom` uses `keyExtractor`.)
